File: morse/morse.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

#define MAX_LEN 2048

typedef struct {
    char ch;
    const char *morse;
} MorseMap;

MorseMap standard_table[] = {
    {'A', ".-"}, 		{'B', "-..."}, 		{'C', "-.-."},		{'D', "-.."},		{'E', "."},
	{'F', "..-."}, 		{'G', "--."}, 		{'H', "...."},		{'I', ".."},		{'J', ".---"},
	{'K', "-.-"}, 		{'L', ".-.."},		{'M', "--"},		{'N', "-."},		{'O', "---"},
	{'P', ".--."},		{'Q', "--.-"},		{'R', ".-."},		{'S', "..."},		{'T', "-"},
    {'U', "..-"}, 		{'V', "...-"}, 		{'W', ".--"}, 		{'X', "-..-"},		{'Y', "-.--"},
	{'Z', "--.."},		{'0', "-----"},		{'1', ".----"},		{'2', "..---"},		{'3', "...--"},
    {'4', "....-"}, 	{'5', "....."},		{'6', "-...."}, 	{'7', "--..."},		{'8', "---.."},
	{'9', "----."},		{'.', ".-.-.-"},	{',', "--..--"},	{'?', "..--.."},	{'\'', ".----."},
    {'!', "-.-.--"},	{'/', "-..-."},		{'(', "-.--."},		{')', "-.--.-"},	{'&', ".-..."},
	{':', "---..."}, 	{';', "-.-.-."}, 	{'=', "-...-"},		{'+', ".-.-."},		{'-', "-....-"},
	{'_', "..--.-"}, 	{'"', ".-..-."},	{'$', "...-..-"}, 	{'@', ".--.-."}, 	{' ', "/"},
	{'\0', NULL}
};

void ascii_to_morse(int ascii, char *buffer) {
    int i;
    strcpy(buffer, "..--");
    for (i = 6; i >= 0; i--) {
        strcat(buffer, (ascii & (1 << i)) ? "-" : ".");
    }
}

const char* encode_char(char c, char *custom_buf) {
    int i;
    for (i = 0; standard_table[i].morse != NULL; i++) {
        if (standard_table[i].ch == c)
            return standard_table[i].morse;
    }
    ascii_to_morse((unsigned char)c, custom_buf);
    return custom_buf;
}

void encode(const char *text, char *output) {
    int i;
    char temp[16];
    output[0] = '\0';
    for (i = 0; text[i] != '\0'; i++) {
        const char *code = encode_char(text[i], temp);
        strcat(output, code);
        strcat(output, "|");
    }
}

char decode_custom(const char *code) {
    if (strncmp(code, "..--", 4) != 0)
        return '?';
    int val = 0;
    for (int i = 4; i < strlen(code); i++) {
        val <<= 1;
        if (code[i] == '-')
            val |= 1;
    }
    return (char)val;
}

char decode_char(const char *code) {
    int i;
    for (i = 0; standard_table[i].morse != NULL; i++) {
        if (strcmp(standard_table[i].morse, code) == 0)
            return standard_table[i].ch;
    }
    return decode_custom(code);
}

void decode(const char *morse, char *output) {
    char buffer[32];
    int i = 0, j = 0, k = 0;
    while (morse[i] != '\0') {
        if (morse[i] == '|') {
            buffer[k] = '\0';
            output[j++] = decode_char(buffer);
            k = 0;
        } else {
            buffer[k++] = morse[i];
        }
        i++;
    }
    output[j] = '\0';
}

void assert_equal(const char *label, const char *expected, const char *actual) {
    if (strcmp(expected, actual) == 0) {
        printf("[PASS] %s\n", label);
    } else {
        printf("[FAIL] %s\n  Expected: %s\n  Actual:   %s\n", label, expected, actual);
    }
}

void run_tests() {
    char encoded[MAX_LEN];
    char decoded[MAX_LEN];

    // Test 1: Basic encoding
    encode("HELLO", encoded);
    assert_equal("Encode HELLO", "....|.|.-..|.-..|---|", encoded);

    // Test 2: Basic decoding
    decode("....|.|.-..|.-..|---|", decoded);
    assert_equal("Decode Morse HELLO", "HELLO", decoded);

    // Test 3: Round-trip
    encode("Test 123!", encoded);
    decode(encoded, decoded);
    assert_equal("Round-trip Test 123!", "TEST 123!", decoded);

    // Test 4: Unknown character
    encode("\x01", encoded); // ASCII 1
    decode(encoded, decoded);
    assert_equal("Unknown ASCII 1", "\x01", decoded);

    // Test 5: Full sentence
    encode("Address: 154 Pensilville - ON, CA", encoded);
    decode(encoded, decoded);
    assert_equal("Full sentence", "ADDRESS: 154 PENSILVILLE - ON, CA", decoded);
}

int main() {
    run_tests();

    // Optional: interactive mode
    char input[MAX_LEN];
    char encoded[MAX_LEN];
    char decoded[MAX_LEN];

    printf("\nEnter a string to encode in Morse code:\n");
    fgets(input, MAX_LEN, stdin);

    size_t len = strlen(input);
    if (len > 0 && input[len - 1] == '\n')
        input[len - 1] = '\0';

    encode(input, encoded);
    printf("\nEncoded Morse:\n%s\n", encoded);

    decode(encoded, decoded);
    printf("\nDecoded Text:\n%s\n", decoded);

    return 0;
}

File: morse/morse.pl
% Morse table facts
morse('A', ".-").     morse('B', "-...").   morse('C', "-.-.").   morse('D', "-..").
morse('E', ".").      morse('F', "..-.").   morse('G', "--.").    morse('H', "....").
morse('I', "..").     morse('J', ".---").   morse('K', "-.-").    morse('L', ".-..").
morse('M', "--").     morse('N', "-.").     morse('O', "---").    morse('P', ".--.").
morse('Q', "--.-").   morse('R', ".-.").    morse('S', "...").    morse('T', "-").
morse('U', "..-").    morse('V', "...-").   morse('W', ".--").    morse('X', "-..-").
morse('Y', "-.--").   morse('Z', "--..").
morse('0', "-----").  morse('1', ".----").  morse('2', "..---").  morse('3', "...--").
morse('4', "....-").  morse('5', ".....").  morse('6', "-....").  morse('7', "--...").
morse('8', "---..").  morse('9', "----.").
morse('.', ".-.-.-"). morse(',', "--..--"). morse('?', "..--.."). morse('\'', ".----.").
morse('!', "-.-.--"). morse('/', "-..-.").  morse('(', "-.--.").  morse(')', "-.--.-").
morse('&', ".-...").  morse(':', "---..."). morse(';', "-.-.-."). morse('=', "-...-").
morse('+', ".-.-.").  morse('-', "-....-"). morse('_', "..--.-"). morse('"', ".-..-.").
morse('$', "...-..-").morse('@', ".--.-."). morse(' ', "/").

% ASCII fallback encoding
ascii_morse(Char, Morse) :-
    char_code(Char, Code),
    Code >= 0, Code =< 127,
    ascii_bits(Code, Bits),
    atom_concat("..--", Bits, Morse).

ascii_bits(Code, Bits) :-
    ascii_bits(Code, 6, "", Bits).

ascii_bits(_, -1, Acc, Acc).
ascii_bits(Code, I, Acc, Bits) :-
    Bit is (Code >> I) /\ 1,
    (Bit =:= 1 -> append_atom(Acc, "-", NewAcc) ; append_atom(Acc, ".", NewAcc)),
    I1 is I - 1,
    ascii_bits(Code, I1, NewAcc, Bits).

append_atom(A, B, C) :- atom_concat(A, B, C).

% Encode a single character
encode_char(Char, Morse) :-
    upcase_atom(Char, Upper),
    (morse(Upper, Morse) -> true ; ascii_morse(Char, Morse)).

% Encode a string
encode_string([], []).
encode_string([Char|Rest], [Morse|EncodedRest]) :-
    encode_char(Char, Morse),
    encode_string(Rest, EncodedRest).

% Decode a Morse symbol
decode_char(Morse, Char) :-
    morse(Char, Morse), !.
decode_char(Morse, Char) :-
    atom_chars(Morse, ['.', '.', '-', '-', |Bits]),
    decode_bits(Bits, 0, Code),
    char_code(Char, Code).

decode_bits([], Val, Val).
decode_bits(['-'|Rest], Acc, Val) :-
    Acc1 is (Acc << 1) + 1,
    decode_bits(Rest, Acc1, Val).
decode_bits(['.'|Rest], Acc, Val) :-
    Acc1 is Acc << 1,
    decode_bits(Rest, Acc1, Val).

% Decode a Morse string
decode_string([], []).
decode_string([Morse|Rest], [Char|DecodedRest]) :-
    decode_char(Morse, Char),
    decode_string(Rest, DecodedRest).

% Utility to convert between strings and char lists
string_to_chars(String, Chars) :- atom_chars(String, Chars).
chars_to_string(Chars, String) :- atom_chars(String, Chars).

% Test runner
run_tests :-
    test("HELLO", "....|.|.-..|.-..|---"),
    test("Test 123!", "TEST 123!"),
    test("\x01", "\x01"),
    test("Address: 154 Pensilville - ON, CA", "ADDRESS: 154 PENSILVILLE - ON, CA").

test(Input, ExpectedOutput) :-
    string_to_chars(Input, Chars),
    encode_string(Chars, MorseList),
    atomic_list_concat(MorseList, '|', MorseAtom),
    atomic_list_concat(MorseList, '|', Encoded),
    split_string(Encoded, "|", "", MorseParts),
    decode_string(MorseParts, DecodedChars),
    chars_to_string(DecodedChars, Decoded),
    (Decoded = ExpectedOutput ->
        format("[PASS] ~w~n", [Input])
    ;
        format("[FAIL] ~w~n  Expected: ~w~n  Actual:   ~w~n", [Input, ExpectedOutput, Decoded])
    ).

% Entry point
:- initialization(main).
main :-
    run_tests,
    write("\nEnter a string to encode in Morse code:\n"),
    read_line_to_string(user_input, Input),
    string_to_chars(Input, Chars),
    encode_string(Chars, MorseList),
    atomic_list_concat(MorseList, '|', MorseAtom),
    format("\nEncoded Morse:\n~w\n", [MorseAtom]),
    split_string(MorseAtom, "|", "", MorseParts),
    decode_string(MorseParts, DecodedChars),
    chars_to_string(DecodedChars, Decoded),
    format("\nDecoded Text:\n~w\n", [Decoded]).
